tycho-common/src/simulation/protocol_sim.rs

```diff
@@ -150,6 +161,117 @@ pub trait ProtocolSim: fmt::Debug + Send + Sync + 'static {
+    /// specified `target_price`. This is used by batch auction solvers to understand the "demand"
+    /// a pool has for token_in at a given price level when constructing optimal execution paths.
+    ///
+    /// The method accounts for the pool's fee structure. The term "demand" reflects the market
+    /// microstructure perspective: from the solver's viewpoint, the pool is expressing demand for
+    /// token_in at this price point.
+    ///
+    /// # Arguments
+    ///
+    /// * `token_in` - The address of the token being sold (swapped into the pool)
+    /// * `token_out` - The address of the token being bought (swapped out of the pool)
+    /// * `target_price` - The maximum acceptable price as a `Price` struct where:
+    ///   - `numerator`: Amount of token_in
+    ///   - `denominator`: Amount of token_out
+    ///   - Represents the price as token_in per token_out (e.g., Price{numerator: 2000,
+    ///     denominator: 1} means willing to pay up to 2000 token_in for 1 token_out)
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(BigUint)` - The maximum amount of token_in that can be traded at or better than the
+    ///   target price. Swapping this amount will result in an average execution price ≤
+    ///   target_price (better or equal from the buyer's perspective).
+    /// * `Err(SimulationError)` - If:
+    ///   - The current pool price is already worse than the target price (no liquidity available at
+    ///     this price point)
+    ///   - The pool has insufficient liquidity to provide meaningful quotes
+    ///   - The calculation encounters numerical issues
+    ///   - The method is not implemented for this protocol
+    ///
+    /// # Notes
+    ///
+    /// - The returned amount includes fees. The actual amount applied to pool reserves will be the
+    ///   gross amount minus protocol fees.
+    /// - This method does not mutate pool state; calculations are read-only
+    /// - The interpretation of "better than target price" depends on trade direction: lower prices
+    ///   favor the buyer of token_out
+    /// - Batch auction solvers use this to aggregate liquidity across multiple pools at specific
+    ///   price levels
+    ///
+    /// # Relationship to swap_to_price
+    ///
+    /// While these methods often return the same value for continuous AMMs, the semantic difference
+    /// is important:
+    /// - `swap_to_price`: "What input moves the marginal price to exactly X?"
+    /// - `query_demand`: "What's the maximum input tradeable at price X or better?"
+    ///
+    /// For protocols with discrete price levels or custom mechanics, these may diverge.
+    #[allow(unused)]
+    fn query_demand(
+        &self,
+        token_in: &Bytes,
+        token_out: &Bytes,
+        target_price: Price,
+    ) -> Result<BigUint, SimulationError> {
+        Err(SimulationError::FatalError("query_demand not implemented".into()))
+    }
```

Comment:
lets rename this to how APEX is right now - query_supply. I think it is a bit clearer than "demand", if this method tells you how much a pool can provide at a given price